backend/regalloc/regalloc_rewrite.ml

```diff
@@ -157,44 +210,46 @@ let rewrite_gen :
           let instr = DLL.value cell in
           if instruction_contains_spilled instr
           then
-            match Regalloc_stack_operands.basic spilled_map instr with
-            | All_spilled_registers_rewritten -> ()
-            | May_still_have_spilled_registers ->
+            if should_coalesce_temp_spills_and_reloads
+               || Regalloc_stack_operands.basic spilled_map instr
+                  = May_still_have_spilled_registers
```

Comment:
Yes, it is necessary to short-circuit here. If the we are coalescing the temporary spills and reloads, we do not want to allow the spilled registers to be directly used in the instruction argument or result arrays.

Reply:
If `should_coalesce_temp_spills_and_reloads` is true at this point, we will definitely coalesce the temporaries in this block as some instruction must contain a spilled register (same for the condition on the terminator). If the block temporary we create gets register allocated, then that is better than using stack operands to use the memory address directly in the instruction. If the block temporary is spilled, stack operand will apply to it in the next round in the same way it would have done to the original variable. So, I think we are fine to short circuit here and will add a comment explaining why.